packages/scripts/src/generate-milestones.ts

```diff
@@ -0,0 +1,175 @@
+  return dayjs(start).add(duration, 'days').format('YYYY-MM-DD');
+};
+
+const addMilestone = async (
+  applicant: IENApplicant,
+  statusId: string,
+  previous: any,
+  job?: IENApplicantJob,
+) => {
+  const milestone: IENApplicantAddStatusAPIDTO = {
+    status: statusId,
+    start_date: getNewStartDate(
+      previous?.start_date || applicant.registration_date || new Date('2021-03-15'),
+    ),
+  };
+  if (job) milestone.job_id = job.id;
+  await axios.post(`/ien/${applicant.id}/status`, milestone);
+  return milestone;
+};
+
+const getAuthorities = async (): Promise<IENHaPcnRO[]> => {
+  const resp = await axios.get('/ienmaster/ha-pcn');
+  return resp?.data?.data;
+};
+
+const createJob = async (applicant: IENApplicant): Promise<IENApplicantJob> => {
+  const authorities = await getAuthorities();
+  const ha = authorities[_.random(0, authorities.length - 1)];
+  const job: IENApplicantJobCreateUpdateDTO = {
+    ha_pcn: ha.id,
+    recruiter_name: faker.name.fullName(),
+    job_post_date: dayjs(applicant.registration_date).add(2, 'months').format('YYYY-MM-DD'),
+  };
+  const resp = await axios.post(`/ien/${applicant.id}/job`, job);
+  return resp?.data?.data;
+};
+
+const generateMilestones = async (id: string, statuses: Record<string, string>) => {
+  const applicant = await getApplicant(id);
+  let previous = undefined;
+  const stages = [
+    LIC_REG_STAGE,
+    NNAS_STAGE,
+    BCCNM_NCAS_STAGE,
+    RECRUITMENT_STAGE,
+    IMMIGRATION_STAGE,
+  ];
+
+  if (applicant.jobs.length) {
+    console.log(`skip applicant ${applicant.id} with a job`);
+    return;
+  }
+
+  const job = await createJob(applicant);
+
+  for (let s = 0; s < stages.length; s++) {
+    for (let i = 0; i < stages[s].length; i++) {
+      const status = stages[s][i];
+      const milestone = applicant.applicant_status_audit.find(s => s.status.status === status);
+      if (milestone) {
+        previous = { start_date: milestone.start_date, status: milestone.status.id };
+        continue;
+      }
+      previous = await addMilestone(
+        applicant,
+        statuses[status],
+        previous,
+        stages[s] === RECRUITMENT_STAGE ? job : undefined,
+      );
+
+      if (status === STATUS.JOB_OFFER_ACCEPTED) break;
+    }
+  }
+};
+
+const setToken = async () => {
+  const token = await getToken();
+  axios.defaults.headers.common['Authorization'] = `Bearer ${token}`;
+  start = new Date();
+  console.log('token updated');
+};
+
+const checkToken = async (start: Date) => {
+  if (dayjs().diff(start, 'minutes') > 3) {
+    await setToken();
+  }
+};
+
+axios.defaults.baseURL = 'http://localhost:4000/api/v1';
+
+setToken()
+  .then(async () => {
+    const [applicants, count] = await getApplicants();
+    const statuses = await getMilestones();
+    for (let i = 0; i < count; i++) {
+      await checkToken(start);
+
+      const applicant = applicants[i];
+      if (dayjs(applicant.registration_date).isAfter('2022-06-01')) {
```

Comment:
comments say `Skip applicants registered before 2022-06-01` should this be isBefore?

Reply:
It generates milestones with random duration, and if the applicant has been registered recently, a milestone date could be a future date.